feat: Add security and privacy pages to the learn section (#107)
ceramicnetwork/docs: docs/learn/advanced/security.md

# Security

This page describes various security aspects of the Ceramic protocol and miscellaneous systems like 3ID Connect.

## Ceramic Protocol

> The main security properties that makes up Ceramic are cryptographic signatures, *proof-of-publication* (though blockchain anchors), and hash-linked data structures. Together these properties allows the construction of a verifiable data structure. In addition to this Ceramic relies on libp2p pubsub to gossip about stream updates.

Reviewer: ```suggestion
The main security properties that makes up Ceramic are cryptographic signatures, *proof-of-publication* (through blockchain anchors), and hash-linked data structures. Together these properties allows the construction of a verifiable data structure. In addition to this Ceramic relies on libp2p pubsub to gossip about stream updates.
```

### Network gossip
Ceramic nodes use libp2p pubsub to gossip about stream tips. The two operations that happen in the gossip is publication of stream updates and queries for tips by nodes that load a specific stream. The main security consideration for a Ceramic node is that any new tip that comes in for a stream which it cares about could be a fake or invalid tip for that stream and thus needs to be verified.

#### DoS attack
A malicious node can spam the pubsub topic by sending a lot of messages. This may be randomly generated messages that are not even for valid streams, or it could be invalid tips for valid streams that get rejected by the stream update rules (the "false log attack" explained below). The main counter attack for this is to limit the amount of messages a single node can send by having an automated reputation system that disconnects from nodes which are spammy. No such reputation systems exists in ceramic today, though it is planned as a future improvement.

#### False log attack

> A malicious node can spam nodes that pin specific streams by sending faulty commit logs. The unsuspecting node would sync the log of the stream and when it validates it the stream would get invalidated and removed from the local node. However, if the false log is long and the malicious node send multiple of these it may cause a significant amount of overhead. The most simple countermeasure to this is to simply stop accepting tips from nodes which have proven to not be reliable. A more significant approach would be to build a StreamType that includes a recursive zero-knowledge proof that the log is indeed correctly associated with the given streamid.

Reviewer: ```suggestion
A malicious node can spam nodes that pin specific streams by sending faulty commit logs. The unsuspecting node would sync the log of the stream and find it invalid, at which point the node would throw out the faulty log without applying it to its local copy of the stream's state. However, if the false log is long and the malicious node sends multiple of these it may cause a significant amount of overhead. The most simple countermeasure to this is to simply stop accepting tips from nodes which have proven to not be reliable. A more significant approach would be to build a StreamType that includes a recursive zero-knowledge proof that the log is indeed correctly associated with the given streamid.
```

Reviewer: Reworded a bit, previous wording made it sound like a faulty log could cause a node to throw out all state for a stream, even the valid state it already had

#### Caip10Link clock synchronization

> Currently the Caip10Link streamtype relies on system time to mitigate replay attacks. A new update to a Caip10Link stream is only valid if the proof includes a timestamp that is larger than the timestamp of the previous update. This presents a problem however if the users system time for some reason is incorrectly far into the future. Users would at this point be unable to prevent a replay attack using a proof created on this machine. A possible workaround could be to switch to using a `nonce` which is always incremented by one, or include the `prev` pointer to the previous commit in the proof.

Reviewer: ```suggestion
Currently the Caip10Link streamtype relies on system time on the client machine to mitigate replay attacks. A new update to a Caip10Link stream is only valid if the proof includes a timestamp that is larger than the timestamp of the previous update. This presents a problem however if the users system time for some reason is incorrectly far into the future. Users would at this point be unable to prevent a replay attack using a proof created on this machine. The replay attack could be used to reset a Caip10Link to point to any DID it had previously been linked to. A possible workaround could be to switch to using a `nonce` which is always incremented by one, or include the `prev` pointer to the previous commit in the proof.
```

> # Security

Reviewer: This page is missing the vulnerability/concerns that were discussed in https://github.com/ceramicnetwork/js-ceramic/pull/666

A *proof-of-publication* is a proof that some content was published at some point in time. This is achieved by publishing the hash of the content on a blockchain. These proofs can be made more cost effective by putting multiple content hashes as leaves in a merkle tree and only publishing the root. The *proof-of-publication* provides a definite ordering of events in a stream, which is useful when you want to do key revocation in a secure manner.

The *linked commit DAG* provides a logical ordering of events which is useful when constructing streams that use CRDT type logic. For a detailed description of these benefits reference the [Merkle-CRDTs](https://research.protocol.ai/blog/2019/a-new-lab-for-resilient-networks-research/PL-TechRep-merkleCRDT-v0.1-Dec30.pdf) paper.

#### Conflict resolution strategies

Different StreamTypes in Ceramic may have different conflict resolution strategies with different security properties.

> The *earliest anchor rule* solves any conflict in the history of a stream by picking the stream that was anchored at the earliest point in time. This enables keys to be securely revoked since someone that gains possession of an old key after it was revoked will be unable to produce a *proof-of-publication* that is earlier than the first anchor. A drawback with this strategy is that the controller of the stream can create anchored commits in secret and reveal them at a later point in time to change the history of the stream. However, this can be mitigated by allowing the log to have multiple branches of history and merge them using a CRDT. This strategy is used by the [TileDocument](../../streamtypes/tile-document/overview.md).

Author: Hm, I guess the data withholding attack is less of an issue in a world where we only accept certain anchor services?

> The *earliest anchor rule* solves any conflict in the history of a stream by picking the stream that was anchored at the earliest point in time. This enables keys to be securely revoked since someone that gains possession of an old key after it was revoked will be unable to produce a *proof-of-publication* that is earlier than the first anchor. A drawback with this strategy is that the controller of the stream can create anchored commits in secret and reveal them at a later point in time to change the history of the stream. However, this can be mitigated by allowing the log to have multiple branches of history and merge them using a CRDT. This strategy is used by the [TileDocument](../../streamtypes/tile-document/overview.md).

Author: I don't see this as being a problem only for 3IDs though. The use of a revoked key could happen for any stream!

> The *earliest anchor rule* solves any conflict in the history of a stream by picking the stream that was anchored at the earliest point in time. This enables keys to be securely revoked since someone that gains possession of an old key after it was revoked will be unable to produce a *proof-of-publication* that is earlier than the first anchor. A drawback with this strategy is that the controller of the stream can create anchored commits in secret and reveal them at a later point in time to change the history of the stream. However, this can be mitigated by allowing the log to have multiple branches of history and merge them using a CRDT. This strategy is used by the [TileDocument](../../streamtypes/tile-document/overview.md).

Reviewer: I don't think any of my comments here were addressed?

My point about 3ids was that the use of blockchain anchoring enables 3ids to support secure key rotation, not that signing messages with revoked keys is only relevant to the 3id stream.  I think we could use this section more to sell 3id and why it's special and so powerful, and how the power of 3id is enabled the security model we chose and built Ceramic around.

Given we're planning to fix the data withholding attack then it's probably fine to not spend too much time on it here.  I actually think we should strike the sentence about using CRDTs to mitigate it as it's not immediately clear how CRDTs help, and that's not even the approach we are taking to fix it.  I also still think the last sentence reads like it's saying that TileDocuments use CRDTs, not that they use Earliest Anchor Wins

> The *earliest anchor rule* solves any conflict in the history of a stream by picking the stream that was anchored at the earliest point in time. This enables keys to be securely revoked since someone that gains possession of an old key after it was revoked will be unable to produce a *proof-of-publication* that is earlier than the first anchor. A drawback with this strategy is that the controller of the stream can create anchored commits in secret and reveal them at a later point in time to change the history of the stream. However, this can be mitigated by allowing the log to have multiple branches of history and merge them using a CRDT. This strategy is used by the [TileDocument](../../streamtypes/tile-document/overview.md).

Author: Thanks for clarifying.

#### False log attack
A malicious node can spam nodes that pin specific streams by sending faulty commit logs. The unsuspecting node would sync the log of the stream and when it validates it the stream would get invalidated and removed from the local node. However, if the false log is long and the malicious node send multiple of these it may cause a significant amount of overhead. The most simple countermeasure to this is to simply stop accepting tips from nodes which have proven to not be reliable. A more significant approach would be to build a StreamType that includes a recursive zero-knowledge proof that the log is indeed correctly associated with the given streamid.

## 3ID Connect

3ID Connect is a key management system for the 3ID DID method. It allows blockchain accounts to be used as an authentication method to the keys used in the users 3ID. In order to authenticate a user signs a message with their wallet. The entropy is used to generate a [Key DID](../../authentication/key-did/method.md) which in turn is used to decrypt the seed used for the 3ID.

> The main security concern here is that the seed used for the 3ID is stored within the 3ID iframe. While this restricts the app using 3ID Connect to do whatever it want's with the seed, a malicious browser plugin could easily pick up the seed from the web browser and take full control over the users 3ID. While are multiple paths that can be used to mitigate this issue, it's currently considered to be a fair tradeoff to improve UX for users. In contrast to a users crypto wallet, the 3ID doesn't actually hold or control any cryptocurrency; it only controls data read/writes. This means that the incentive for a hacker to execute an attack is much smaller since there is no immediate financial reward. Also note that this security issue is not inherent to 3ID itself, it's just related to how 3ID is used within 3ID connect.

Reviewer: ```suggestion
The main security concern here is that the seed used for the 3ID is stored within the 3ID iframe. While this restricts the app using 3ID Connect from doing whatever it wants with the seed, a malicious browser plugin could easily pick up the seed from the web browser and take full control over the users 3ID. While are multiple paths that can be used to mitigate this issue, it's currently considered to be a fair tradeoff to improve UX for users. In contrast to a users crypto wallet, the 3ID doesn't actually hold or control any cryptocurrency; it only controls data read/writes. This means that the incentive for a hacker to execute an attack is much smaller since there is no immediate financial reward. Also note that this security issue is not inherent to 3ID itself, it's just related to how 3ID is used within 3ID connect.
```

Reviewer: Wow I didn't know that about the malicious browser extension being able to take over users' DIDs.  That seems pretty scary

Reviewer 2: This attack vector is not specific to 3id frame. It is known for ages, that a browser extension can steal cookies or credit card details easily.

Reviewer: ah yeah, I guess browser extensions can always kind of do whatever they want. I retract my surprise/alarm.

### Possible security improvements

#### External 3ID controller

> The 3ID did document has different public keys used for writing and controlling the DID document itself. One possible mitigation strategy is to store the controller key in a more secure manner. This would make it possible to recover if the day to day writing public keys are compromised. An example solution here could be to store the controller key in a MetaMask Snap.

Reviewer: ```suggestion
The 3ID DID document has different public keys used for writing and controlling the DID document itself. One possible mitigation strategy is to store the controller key in a more secure manner. This would make it possible to recover if the day to day writing public keys are compromised. An example solution here could be to store the controller key in a MetaMask Snap.
```

Reviewer: "One possible mitigation strategy" - mitigation for what?  Weird to word it this way using the word "mitigation" when you haven't really laid out the problem/vulnerability yet

Author: Good catch, ty.

Different StreamTypes in Ceramic may have different conflict resolution strategies with different security properties.

The *earliest anchor rule* solves any conflict in the history of a stream by picking the stream that was anchored at the earliest point in time. This enables keys to be securely revoked since someone that gains possession of an old key after it was revoked will be unable to produce a *proof-of-publication* that is earlier than the first anchor. This is used in the [3ID DID Method](../../authentication/3id-did/method.md) which registers a set of public keys associated to a DID URL, when the 3ID stream is updated with new public keys, revoking the old keys, an outside observer can be sure of the state of the 3ID without any secondary sources of information. 3IDs are built on top of the [TileDocument](../../streamtypes/tile-document/overview.md) stream type, and alternative DID methods could also be built on top of it.

> A drawback with this strategy is that the controller of the stream can create anchored commits in secret and reveal them at a later point in time to change the history of the stream. There are various ways to mitigate this. One of them is to only allow anchor services that actively make sure that the update is valid before it's being anchored.

Reviewer: ```suggestion
A drawback with the "earliest anchor rule" is that the controller of the stream can create anchored commits in secret and reveal them at a later point in time to change the history of the stream. There are various ways to mitigate this. One of them is to only allow anchor services that actively publish all created anchor commits when they are created.
```

Reviewer: the issue is keeping the commits secret, not that they were invalid at the time that they were created, right?
